crates/sui-node/src/lib.rs

```diff
@@ -200,68 +202,87 @@ impl SuiNode {
+                let consensus_adapter = consensus_adapter.clone();
+                join_set.spawn(
+                    async move {
+                        // note: restart-safe de-duplication happens after consensus, this is
+                        // just best-effort to reduce unneeded submissions.
+                        let mut seen = HashSet::new();
+                        loop {
+                            info!("fetching JWK for provider {:?}", p);
+                            match Self::fetch_jwks(&p).await {
+                                Err(e) => {
+                                    warn!("Error when fetching JWK {:?}", e);
+                                    // Retry in 30 seconds
+                                    tokio::time::sleep(Duration::from_secs(30)).await;
+                                }
+                                Ok(mut keys) => {
+                                    keys.retain(|(id, _)| !seen.insert(id.clone()));
+
+                                    // prevent oauth providers from sending too many keys,
+                                    // inadvertently or otherwise
+                                    if keys.len() > MAX_JWK_KEYS_PER_FETCH {
+                                        warn!("Provider {:?} sent too many JWKs, only the first {} will be used", p, MAX_JWK_KEYS_PER_FETCH);
+                                        keys.truncate(MAX_JWK_KEYS_PER_FETCH);
+                                    }
+
+                                    for (id, jwk) in keys.into_iter() {
+                                        info!("Submitting JWK to consensus: {:?}", id);
+
+                                        let txn = ConsensusTransaction::new_jwk_fetched(id, jwk);
+                                        consensus_adapter.submit(txn, None, &epoch_store)
+                                            .tap_err(|e| warn!("Error when submitting JWKs to consensus {:?}", e))
+                                            .ok();
+                                    }
                                 }
                             }
-                        }
-                        // Sleep for 1 hour
-                        tokio::time::sleep(Duration::from_secs(3600)).await;
-                    };
-
-                    tokio::select! {
-                        _ = fetch_and_sleep => {}
-                        _ = epoch_store.wait_epoch_terminated() => {
-                            break;
+                            // Sleep for 1 hour
+                            tokio::time::sleep(Duration::from_secs(3600)).await;
                         }
                     }
-                }
-                info!("JWK updater task terminated");
+                    .instrument(error_span!("jwk_updater_task", epoch)),
+                );
             }
-            .instrument(error_span!("jwk_updater_task", epoch)),
-        );
+
+            epoch_store.wait_epoch_terminated().await;
```

Comment:
Some nit picking on the tasks - right now we send stop signal to task when epoch ends, but tasks can still continue running while epoch end happens, as we don't wait for the to stop.
This is probably ~ok, but perhaps we should just add a "barrier" here, to make sure those tasks stop before epoch ends, just so that we avoid questions like "can something go wrong".

We have a convenient function for this - `AuthorityEpochStore::within_alive_epoch` - it drops task when epoch ends and waits for it to complete before proceeding with reconfiguration

Reply:
good idea, will add that. I can probably get rid of the join set then